[MXNET-1404] Added the GPU memory profiler

ArmageddonKnight opened the pull request:
## Description ##
Added the GPU memory profiler.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] In the backend, added an extra name `tag` to the `Storage::Handle`, `NDArray` and `Resource` classes, and a new profiler `GpuMemoryProfiler` that records GPU memory allocations and store them to file.
- [x] In the Python bindings, added an extra `name` slot to the `NDArrayBase` class<del>, and modified the Gluon symbol names to be more specific</del>.
- [x] Finally, added an example that include instructions on how to use the GPU memory profiler and scripts that help visualization and analyzing.

![MXNet-GPU_Memory_Profiler-Design](https://github.com/dmlc/web-data/blob/master/mxnet/example/gpu_memory_profiler/MXNet-GPU_Memory_Profiler-Design.png?raw=true)

## Comments ##
The GPU memory profiler was demoed at the **SysML 2019** ([Demo #24](https://www.sysml.cc/doc/2019/demo_24.pdf)). The demo video (credit to Geoffrey Yu @geoffxy) is attached to this pull request. Please kindly check the video for expected profiling behavior. Also, please note that we have made several changes since the demo so the current version (this pull request) might look slightly different.

![MXNet-GPU_Memory_Profiler-S](https://github.com/dmlc/web-data/blob/master/mxnet/example/gpu_memory_profiler/MXNet-GPU_Memory_Profiler.gif?raw=true)

szha commented:
Awesome work!

ArmageddonKnight commented:
@szha Thanks. Any feedback or suggestions are welcome.

eric-haibin-lin commented on python/mxnet/optimizer/optimizer.py:
```diff
@@ -581,7 +581,9 @@ def create_state(self, index, weight):
             stype = weight.stype if self.lazy_update else 'default'
-            momentum = zeros(weight.shape, weight.context, dtype=weight.dtype, stype=stype)
+            momentum = zeros(weight.shape, weight.context, dtype=weight.dtype, stype=stype,
+                             name=weight.name.replace("in_arg", "optimizer_state:momentum"))
```
This seems a bit adhoc

ArmageddonKnight replied:
Please let us know if there is any better way of annotating those optimizer states.

eric-haibin-lin commented on src/profiler/gpu_memory_profiler.cc:
```diff
@@ -0,0 +1,103 @@
+}
+
+GpuMemoryProfiler::GpuMemoryProfiler() {
+  const char * use_gpu_memory_profiler = getenv("MXNET_USE_GPU_MEMORY_PROFILER");
```
Use `dmlc::GetEnv` instead of getenv

eric-haibin-lin commented on example/gpu_memory_profiler/README.md:
```diff
@@ -0,0 +1,151 @@
+    which controls the storage tagging (disabled by default),
+    **and** `MXNET_USE_GPU_MEMORY_PROFILER` flag in `src/profiler/gpu_memory_profiler.h` **1**,
+    which controls dumping the GPU memory allocation inforamtion (also disabled by default).
+  Please note that you must have **BOTH flags** set to **1** to use the GPU memory profiler.
```
Is there any use case where developers only want to enable 1 of them? Why not have 1 flag for turning on/off this feature for simplicity?

ArmageddonKnight replied:
`MXNET_ENABLE_STORAGE_TAGGING` is used for tagging `StorageHandle`'s while `MXNET_USE_GPU_MEMORY_PROFILER` is used for dumping. They serve for different purposes so that's why I keep them separate.

eric-haibin-lin commented on src/operator/rnn-inl.h:
```diff
@@ -1378,9 +1378,17 @@ class RNNOp {
       workspace_size_ = workspace_byte_ / sizeof(DType);
       // Allocate the reserve space
-      reserve_space_ = Storage::Get()->Alloc(reserve_space_byte_, Context::GPU(s->dev_id));
+      reserve_space_ = Storage::Get()->Alloc(reserve_space_byte_, Context::GPU(s->dev_id)
```
Did you measure the overhead of always passing the string? Having these macros is not easy to maintain. 
If the overhead is neglegible it's probably cleaner to alway pass the name

ArmageddonKnight replied:
In my opinion the overhead is very difficult to measure. The reason why I added those flags is because there might be cases where people just want to work with a *clean* version of MXNet with nothing added and hence do not need to worry about the overhead.

ArmageddonKnight commented on example/gpu_memory_profiler/SETME.py:
```diff
@@ -0,0 +1,40 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+""" SETME.py
```
It means that users should set the variables within this file (the keyword dictionaries and path to the memory profile) before using the GPU memory plotter and analyzer,

ArmageddonKnight commented:
@anirudh2290 Thanks for your comment. Given below is my thought when developing this memory profiler.

1. I have included the high-level design ideas as one of the examples `./example/gpu_memory_profiler/README.md`. Please let me know if you think more details are needed.
2. I am not using the existing Python profiler API because the memory usage profiling is very different from the VTune-based CPU profiling or NVTX-based GPU profiling. For instance, it does not really make sense to say `pause` or `resume` in the memory profiling because you always care about the total memory consumption rather than some portion of your application. Furthermore, since the format of dump files (`.csv` vs. `.json`) and the visualization tools (`matplotlib` vs. chrome tracing) are completely different, I do not see a very decent way of integrating the GPU memory profiler into the current profiler API.
3. Those environmental variables are added to give users fine-grained control on the path and name of the output files. Since they both have default values, users do not necessarily need to set them explicitly.
4. To my best knowledge, we cannot avoid adding those APIs, because otherwise there is no such a *path* to propagate the name tags from the Python frontend to the C++ backend.
5. Those build flags are needed to switch on and off the GPU memory profiler. They are different from the current `USE_PROFILER` build flag since they are targeting at the GPU memory consumption instead of performance. I have been doing experiments using the memory profiler and so far **do not see any performance degradation** (after all, in most cases the memory allocations are only done once for the entire training process). It is really hard to argue about the exact runtime overhead since the GPU memory profiling requires changes on both the Python frontend and the C++ backend. However, considering that there might be people who just want to work with a "clean" version of MXNet, I added those compilation flags and default them to off.

larroy commented on example/gpu_memory_profiler/README.md:
````diff
@@ -0,0 +1,151 @@
+- Build the MXNet core library.
+
+```bash
+cd mxnet # workspace/mxnet
````
Can we refer to the install instructions? These will quickly become obsolete. For example now it indicates make vs CMake, maybe is better to just say follow the build linked here and add this build flag...

larroy commented on example/gpu_memory_profiler/README.md:
````diff
@@ -0,0 +1,151 @@
+
+```bash
+cd python # workspace/mxnet/python
+python3 setup.py install
````
I think this is duplicating the build & install instructions.  @aaronmarkham  what do you think?

larroy commented on example/gpu_memory_profiler/README.md:
```diff
@@ -0,0 +1,151 @@
+
+## !Attention
+
+*As you use the GPU memory profiler, please kindly pay close attention to the followings.*
```
s/followings/following/

larroy commented on example/gpu_memory_profiler/README.md:
```diff
@@ -0,0 +1,151 @@
+- The information reported by the GPU memory profiler depends on the symbol names
+    you assign at the Python frontend.
+  Therefore, it is important that you give meaningful names as you program your models.
+- It is **STRONGLY RECOMMENDED** that you use the **NAIVE** engine and storage pool for the GPU memory profiling
```
Should we set it ourselves inside the engine? GPU_PROFILE => NAIVE ENGINE

ArmageddonKnight replied:
The GPU memory profiler also works for non-naive engine types. The problem with non-naive engines is that they can make the GPU memory profiling logs out-of-order and hence hard to read and digest. This is why I added it as a suggestion rather than a strong restriction.

larroy replied:
Thanks for the clarification

larroy commented on example/gpu_memory_profiler/SETME.py:
```diff
@@ -0,0 +1,40 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Could you add python3 shebang?

larroy commented on example/gpu_memory_profiler/pprint_top_entries.py:
```diff
@@ -0,0 +1,51 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
py3 shebang

larroy commented on example/gpu_memory_profiler/profile_analyzer.py:
```diff
@@ -0,0 +1,103 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
same

larroy commented on example/gpu_memory_profiler/profile_analyzer.py:
```diff
@@ -0,0 +1,103 @@
+
+                kw_matched = None
+
+                for key, val in kw_dict.items():
```
could we separate this heavily nested code into smaller functions?

larroy commented on example/gpu_memory_profiler/profile_analyzer.py:
```diff
@@ -0,0 +1,103 @@
+                        if kw in tag:
+                            if kw_matched is not None:
+                                # pylint: disable=logging-not-lazy
+                                logging.warning("Tag %s is considered match for "
```
you can directly pass the strings as additional arguments to the logging function, no need to format since using %s already.

larroy commented on example/gpu_memory_profiler/two_bar_plot.py:
```diff
@@ -0,0 +1,118 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
shebang

larroy commented on include/mxnet/c_api.h:
```diff
@@ -521,6 +521,26 @@ MXNET_DLL int MXNDArrayCreate(const mx_uint *shape,
                               NDArrayHandle *out);
 
+/*!
+ * \brief create a NDArray with specified shape and name tag
```
nice!

larroy commented on include/mxnet/ndarray.h:
```diff
@@ -102,15 +112,29 @@ class NDArray {
           std::vector<int> aux_types = {}, mxnet::ShapeVector aux_shapes = {},
-          mxnet::TShape storage_shape = mxnet::TShape(mshadow::Shape1(0)));
+          mxnet::TShape storage_shape = mxnet::TShape(mshadow::Shape1(0))
+#if MXNET_ENABLE_STORAGE_TAGGING
```
Why not always have this enabled and not have the define? memory concerns?

ArmageddonKnight replied:
Mostly performance concerns. Since the GPU memory profiler touches both the Python frontend and the C++ backend, it is hard to measure the actual overhead of propagating the name tags all the way down from the Python APIs to the C++ storage handles. Hence, this flag is added for people who just want to work with a "clean" version of MXNet.

larroy commented on example/gpu_memory_profiler/README.md:
````diff
@@ -0,0 +1,151 @@
+sudo apt-get install ...
+```
+
+- Modify the `MXNET_ENABLE_STORAGE_TAGGING` flag in `include/storage_tag.h` to **1**,
````
It should be exposed through the build system as a define I think

ArmageddonKnight replied:
I previously also considered making those flags as part of the build system but eventually decided to put them inside the source files. The reason is because if we put them inside the build system then every time people want to turn on/off the profiler they have to compile the entire codebase from source which can be really tedious. On the contrary, putting them inside the source files allow the build system to automatically figure out the dependencies and recompile only the files that are affected.

ArmageddonKnight commented:
@larroy That would be great. I took out the large `.gif` file from the pull request but I would really appreciate if we could have it in the MXNet channel. The `.gif` file was actually compressed and converted from the original `.mp4` video and hence is in low-resolution and does not give people the ability of pausing to learn more details. Please let us know if there is any way we could forward the original video to you.

Also, thank you so much for your valuable feedbacks. I think I have addressed most of them

eric-haibin-lin commented:
The gif file is quite helpful, what about uploading to https://github.com/dmlc/web-data and add link it

ArmageddonKnight commented:
@eric-haibin-lin An excellent suggestion. I have opened the pull request at the `dmlc/web-data` repository (https://github.com/dmlc/web-data/pull/175). Please kindly accept it. Thanks.

szha commented:
@ArmageddonKnight done

ArmageddonKnight commented:
@szha Wow. That's quick. Thanks a lot.

pinaraws commented:
@mxnet-label-bot add[Profiler, pr-awaiting-response]

anirudh2290 commented on example/gpu_memory_profiler/README.md:
```diff
@@ -0,0 +1,146 @@
+
+- Modify the `MXNET_ENABLE_STORAGE_TAGGING` flag in `include/storage_tag.h` to **1**,
+    which controls the storage tagging (disabled by default),
+    **and** `MXNET_USE_GPU_MEMORY_PROFILER` flag in `src/profiler/gpu_memory_profiler.h` **1**,
```
I tried this. I was able to generate csv and log. But I wasnt able to use the analyzer. I think I am using SETME wrongly. Some examples in SETME would help.

ArmageddonKnight replied:
I agree that adding a microbenchmark might be helpful in this case. Added to the TODO list.

anirudh2290 commented on example/gpu_memory_profiler/README.md:
```diff
@@ -0,0 +1,146 @@
+- The GPU memory profiler works for Gluon `HybridBlock`s as well, as it also implicitly creates 
+    symbolic computation graphs under the hood after `hybridize()` activation.
+  However, it **DOES NOT WORK WELL** with the imperative programming paradigm 
+    (i.e., `NDArray` function calls).
```
would be great to have an example each of symbol and gluon hybrid block.

anirudh2290 commented on example/gpu_memory_profiler/SETME.py:
```diff
@@ -0,0 +1,45 @@
+        "Parameters"   : ["in_arg", "arg_grad", "optimizer_state"],
+        "Data Entries" : ["data_entry"],
+        "Workspace"    : ["workspace"]
+    }
```
Please provide some examples of how to use in comments.

anirudh2290 commented on include/mxnet/c_api.h:
```diff
@@ -534,13 +554,34 @@ MXNET_DLL int MXNDArrayCreate(const mx_uint *shape,
+                                     int delay_alloc,
+                                     int dtype,
+                                     const char *name,
+                                     NDArrayHandle *out);
```
I didnt see this use anywhere. Where is this called from ?

ArmageddonKnight replied:
It is called by the Python frontend API. In `ndarray/ndarray.py`, the original call to `MXNDArrayCreateEx` was replaced by `MXNDArrayCreateExWName`

anirudh2290 commented on include/mxnet/ndarray.h:
```diff
@@ -90,10 +90,20 @@ class NDArray {
+          int dtype = mshadow::default_type_flag
+#if MXNET_ENABLE_STORAGE_TAGGING
+        , const std::string& name = MXNET_DEFAULT_STORAGE_TAG("unknown")
+#endif  // MXNET_ENABLE_STORAGE_TAGGING
```
can we just default as nullptr and use #if inside the function definition. same at other places.

ArmageddonKnight replied:
The reason why the default name is set to the `MXNET_DEFAULT_STORAGE_TAG` instead of `""` is because `MXNET_DEFAULT_STORAGE_TAG` defaults the string value to the **function name of the caller**.

This is important for avoiding making intrusive changes inside the operator implementations. For instance, as we have shown above, each MXNet operator can potentially request temporary resource, without the `MXNET_DEFAULT_STORAGE_TAG` macro that allows us to trace back to the caller, we need to annotate the resource allocations inside each operator, which is obviously not a very scaleable solution.

anirudh2290 commented on include/mxnet/ndarray.h:
```diff
@@ -1071,7 +1227,11 @@ class NDArray {
+        aux_handles[i] = Storage::Get()->Alloc(aux_bytes, ctx
+#if MXNET_ENABLE_STORAGE_TAGGING
+          , shandle.tag + ":" + AuxIdxToString(i)
+#endif  // MXNET_ENABLE_STORAGE_TAGGING
```
can we avoid this preprocessor directive between a statement. To elaborate, change this to:
```
# if MXNET_ENABLE_STORAGE_TAGGING
aux_handles[i] = Storage::Get()->Alloc(aux_bytes, ctx, shandle.tag + ":" + AuxIdxToString(i));
#else 
aux_handles[i] = Storage::Get()->Alloc(aux_bytes, ctx);
#endif

```

ArmageddonKnight replied:
I previously also tried to avoid adding preprocessor directive to between a statement to avoid adding the `NOLINT(*)`statement. However, for functions that have a long list of arguments, this can be a nightmare. It also makes it harder to maintain as you always need to modify both code segments.

ArmageddonKnight commented:
Hi @anirudh2290 ,

Thanks for your valuable feedback. Let me record a list of Discussion and TODO items:

## Discussion

- [ ] Preprocessor Directives (1. 2.). 
- [ ]
- [ ] Imperative Support (4.). We should **NOT** drop the support for imperative, as this will significantly boost the amount of GPU memory allocated with the unknown tag. For instance, currently most optimizer states are allocated using the pure imperative approach. In fact, almost all the current optimizer implementations (e.g., SGD, Adam) initialize the optimizer states with `mx.nd.zero`. If we drop the imperative support, then all of those allocations will fall to the `unknown` category, which can be `1 GB` for large models.
- [ ] Profiler API Integration (5. 6. 7. 8.). The GPU memory profiler is different from the existing profilers in many ways: (1) It is not using the `chrome://tracing` visualization backend, and the reason is because it needs to accept the users' input on defining the **keyword dictionaries for grouping storage tags** (also, I do not see a very good way of visualizing percentage using `chrome://tracing`). (2) Because it 

## TODO 

- [ ] Add minimum working example in the `example` directory to show how `SETME`, analyzer, and plotter work.
- [ ]

szha commented:
We had a discussion yesterday about the design (w/ @ArmageddonKnight  @anirudh2290 @eric-haibin-lin). We all agreed that this is an awesome feature to be added to mxnet. There were some concerns around the current design and one particular point is about requiring names passed from the frontend for each NDArray, along with the addition of the named imperative invoke interface. The motivation was to be able to identify and track the source of allocation to specific code.

As agreed, here I offer an alternative design (and mock-up for the user experience) for automatically naming NDArrays that are easy to identify, without requiring manual naming of all NDArrays. It can be done by providing the interfaces to make it easy for users to mark a region of code. That, in combination with the information from the operators, should allow users to easily identify the part of code that's responsible for certain memory allocation.

The design enables:
- user specified scope in code. 
- identifying arrays within that scope.
- without changes to existing execution interface like imperative invoke.

Suppose user wants to profile a function that looks like:
```
def function1(nd1, nd2):
    r1 = mx.nd.op1(nd1)
    r2 = mx.nd.op1(nd2)
    r3 = mx.nd.op2(r1, r2)
    return r3
```

One easy way is to have an interface to the users for marking the scope.
```
def function1(nd1, nd2):
    with mx.profiler.scope('function1'):
        r1 = mx.nd.op1(nd1)
        r2 = mx.nd.op1(nd2)
        r3 = mx.nd.op2(r1, r2)
        return r3
```

When entering the profiler scope, it can invoke a new C API that sets the thread local context of the profiler scope name by prepending the outer scope name (if any) to the user specified scope name, and saves the old scope. When exiting, it restores the old scope. When allocation happens, the entries should be named according the current scope name, and have a counter-based naming for op to differentiate multiple invocation of the same op (i.e. the allocation identifier should look like `{scope}.{op}.{counter}`). From the code example, this should have the effect of producing these allocation records:
```
name,bytes
function1.op1.1,x
function1.op1.2,y
function1.op2.1,z
```

If user wants to have more granularity in the scope of the code, this is possible:
```
def function1(nd1, nd2):
    with mx.profiler.scope('function1'):
        r1 = mx.nd.op1(nd1)
        r2 = mx.nd.op1(nd2)
        with mx.profiler.scope('sec1'):
            r3 = mx.nd.op2(r1, r2)
        return r3
```
which results in the following records:
```
name,bytes
function1.op1.1,x
function1.op1.2,y
function1.sec1.op2.1,z
```

For easier usage, we can also utilize the decorator syntax:
```
@mx.profiler.scope('function1')
def function1(nd1, nd2):
    r1 = mx.nd.op1(nd1)
    r2 = mx.nd.op1(nd2)
    r3 = mx.nd.op2(r1, r2)
    return r3
```
The decorator should wrap the decorated code within the aforementioned scope.

For Gluon, we can update the `__call__` function in `Block` and `HybridBlock` and automatically use the class names as scopes. For optimizers, we can update the `update` function to achieve similar effect.

Also, notice that the allocation identifier already has a hierarchy that can be used for aggregation. This can be utilized to aggregate results without requiring the mapping from the users. This means there would no longer be the need to ask for `SETME.py` changes.

This design can be applied to both Symbol and NDArray. The allocation identifier should be treated separately from the variable names in Symbol, so that it provides a consistent experience for Gluon HybridBlocks.

ArmageddonKnight commented:
@szha The problem with this proposed approach is that it might work well for pure imperative programming paradigm, but **NOT** for symbolic graphs or Gluon.

The reason is because in those approaches memory allocations do not happen immediately. They happen only when the computation graphs are materialized. Consider the same piece of code, but with the symbolic graph approach:

```Python
def function1(sym1, sym2):
    with mx.profiler.scope('function1'):
        r1 = mx.sym.op1(sym1)
        r2 = mx.sym.op1(sym2)
        r3 = mx.sym.op2(r1, r2)
        return r3
```

Because `mx.sym` only builds up the graph but does not do actual memory allocations or compute, by the time when the actual GPU memory allocations happen (i.e., when the computation graphs are materialized using `bind()` or `simple_bind()`), the profiler scope information is already lost and I do not think there is a very good way of recovering such information.

Besides, with your proposed approach, all the existing MXNet implementations need to be modified with the `mx.profiler.scope('...')` injected if they want to use the memory profiler. Given that frontend programmers always need to provide extra annotations, why not put them all inside one file rather than letting them scattering around in the source code?

szha commented:
@ArmageddonKnight the allocation identifier in this approach can be added to symbols in the exact same way as NDArrays, and it should happen during symbol construction.

I'm not sure I follow the argument about the code being "scattered around". Also, I honestly don't see any way for other approaches to be more "scattered around" than naming every NDArray.

ArmageddonKnight commented:
@szha But this means that we need to change the `nnvm` codebase, since `nnvm::Graph` is the one that is responsible for managing those symbols. We also need to somehow propagating this scope information from the forward computation node to its corresponding backward one. Such things require changes that are completely different from what we are doing now.

Let me summarize my primary concerns with the proposed solution.

1. It requires **more code changes**, both in terms of **breadth & depth**.
  - In terms of breadth, **NONE** of the existing Python frontend API (e.g., [`LSTMCell`](https://github.com/apache/incubator-mxnet/blob/01cf29d0aaf8af7424fa82d8100cbb967860c712/python/mxnet/rnn/rnn_cell.py#L408)) and applications (e.g., [**Sockeye**](https://github.com/awslabs/sockeye), [**MXNet Examples**](https://github.com/apache/incubator-mxnet/tree/master/example)) are having this profiler scope information, which means that they ALL need to be modified with such information added.
  - In terms of depth, because the GPU memory allocations are delayed for symbolic graphs until they are materialized. `nnvm::Graph` now needs to be aware of the profiler scope information, which means that we need to modify the *nnvm* codebase to achieve this.
2. **Automatic aggregation** can be **undesirable**.
  - I do admit that automatic aggregation seems really cool, and I think this is the currently adopted approach by the *Tensorflow* GPU memory profiler. However, there are problems with this approach:
    - "*The GPU memory profiler tells me that the attention layers consume 50% of the memory footprint. Now what?*" As we group symbols together into profiler scope, we lose the information of the memory consumption on each individual symbol. Although this information can be recovered by gradually refining the profiler scope, it also requires multiple run of the application to find the appropriate granularity, which can be really tedious for applications such as [**Sockeye**](https://github.com/awslabs/sockeye) that spend large amount of time in data preprocessing.
    - This is why I suggest we should instead go for the "*measure once, aggregate multiple times*" approach. We start with the big picture on where the most of the memory consumptions go, and then gradually focus ourselves on the exact symbol(s) that are causing trouble.

To conclude, the table below shows the comparison

|  | Proposed Solution | Current Solution |
| --- | --- | --- |
| Code Changes | All existing Application Codebase | `SETME.py` (usually 3~5 lines of code) |
| Profiler Scope | Fixed. Need to rerun the application for a more coarse or fine-grained profiler scope.   |  Flexible, keyword dictionary can be redefined multiple times in `SETME.py`. No need to rerun the application since the profiler log is present. |

szha commented:
> NONE of the existing Python frontend API (e.g., LSTMCell) ... are having this profiler scope information

As mentioned before, for Gluon Blocks such as LSTMCell, we can use the class name as the default scopes so that users don't need to manually mark them.

> and applications (e.g., Sockeye, MXNet Examples)

I assume this points to the symbolic API examples. I think it's safe to not worry too much about that for now, because 1) we recommend users to use HybridBlocks instead of using symbolic API directly. 2) code base such as sockeye are migrating to Gluon API.

> In terms of depth ... we need to modify the nnvm codebase to achieve this

Even if we are to implement this for symbolic API, setting additional attribute in MXNet C API should suffice for recording the identifiers without any change to nnvm. If solely rely on symbol names, there's no guarantee that the name is consciously chosen or recognizable to the users, as it's common practice to only name the input variables.

> Although this information can be recovered by gradually refining the profiler scope, it also requires multiple run of the application to find the appropriate granularity, which can be really tedious for applications such as Sockeye that spend large amount of time in data preprocessing.

Given that the data preprocessing involves mostly NDArray, using scope be favorable compared to naming every NDArray. It's `O(# of scopes)` vs `O(# of arrays)` even if comparing just the manual work.

> This is why I suggest we should instead go for the "measure once, aggregate multiple times" approach.

The alternative approach for naming the identifier makes each variable uniquely identifiable, which is what can be achieved in the current PR's approach only when users name the variables carefully. Also, the hierarchy information in the profiler dump gives strictly more information about the structure of the program. Finally, there's nothing preventing users from changing the aggregation level if that's desirable, and more intelligent options should be possible rather than asking for the mapping for each name from the users.

And finally, based on the answers above, here's how I would compare these approaches:

|  | Proposed Solution | Current Solution |
| --- | --- | --- |
| Code Changes | All existing Application Codebase | ~`SETME.py` (usually 3-5 lines of code)~<br/> - all imperative code bases on the order of `O(# of arrays)`.<br/> - plus manual mapping of all array names for different aggregation levels, again on the order of `O(# of arrays)` |
| Profiler Scope | ~Fixed. Need to rerun the application for a more coarse or fine-grained profiler scope.~<br/> - no need to rerun. the only thing changed was how the names are assigned, which doesn't prevent logging.<br/>- hierarchical information is feasible and already exists in Gluon blocks.<br/> - different aggregation levels can be used if so desired. |  Flexible, keyword dictionary can be redefined multiple times in `SETME.py`. No need to rerun the application since the profiler log is present. (@szha: yes, at the cost of `O(# of array)` both in terms of labor and in terms of human memory.) |

ArmageddonKnight commented:
@szha I am afraid that there is some serious misunderstanding going on. For all the MXNet benchmarks we have examined so far, none of them require <a href="https://www.codecogs.com/eqnedit.php?latex=O(\text{Number&space;of&space;Arrays})" target="_blank"><img src="https://latex.codecogs.com/gif.latex?O(\text{Number&space;of&space;Arrays})" title="O(\text{Number of Arrays})" height="15pt" /></a> lines of code changes to work.

For the LSTM-based machine translation and BERT, we made no changes to the original source code. Simply adding the following lines to `SETME.py` suffice to complete the GPU memory profiling job.

```Python
# LSTM-based Machine Translation
layer_kw_dict = \
    {
        "RNN"       : [ "rnn" ],
        "Attention" : [ "att" ],
        "Embed"     : [ "embed" ],
        "Output"    : [ "softmax", "logits" ]
    }
```

```Python
# BERT
layer_kw_dict = \
    {
        "Embed"           : [ "embed" ],
        "AttentionCell"   : [  "multiheadattentioncell",
                              "dotproductattentioncell" ],
        "PositionwiseFFN" : [ "bertpositionwiseffn" ]
    }
```

I do not see why you keep saying it requires changing "number of arrays" lines of code for the GPU memory profiler to work.

szha commented:
This is because the current approach identifies the allocation source by naming every NDArray from the frontend. The quoted examples in GluonNLP happen to use `name_scope` and thus assigns names in symbolic mode. Since mxnet 1.3 we no longer require users to use `name_scope`, which means we cannot rely on it for getting meaningful names. Thus, for general cases, users need to assign names manually for each array.

anirudh2290 commented:
> ## Discussion
> * [x]  Preprocessor Directives (1. 2.).
>   
>   * The reason why I included the GPU memory profiler as header flags rather than compilation flags is because in the former implementation, when users decide to switch on/off the GPU memory profiler, **the build system can automatically help them figure out the build dependency**, whereas in the latter implementation they need to compile the entire MXNet from scratch.
> * [x]  Separate Preprocessor Directives (3.).
>   
>   * As their names indicate, **storage tagging** and **GPU memory profiling** are actually two separate things - **storage tagging** is equally applicable to CPU memory profiling (or other devices) while **GPU memory profiling** is only an example that benefits from it. That is why I keep them separate.
>   * I would like to propose **PERMANENTLY adding storage tagging** to the C++ backend, as I believe it might be a useful feature in the future and I do not think propagating an extra string from the Python frontend to the C++ backend will hurt performance because: (1) most strings are small (2) the storage tagging only happens once at the beginning of the training.

I think we can have additional tag parameter in ndarray, storage and resource. I think it should be fine and I dont see this additional parameter for these different API calls becoming a bottleneck.

I think we should do this PR in an incremental fashion though. First phase should just allow for API changes limited to MXNET_ENABLE_STORAGE_TAGGING. Once we get that merged, we can monitor our internal benchmarks for change in performance and move on to the additional changes. I am open to alternative suggestions.

> * [ ]  Profiler API Integration (5. 6. 7. 8.)
>   
>   * The GPU memory profiler is different from the existing profilers in many ways:
>     (1) It is not using the `chrome://tracing` visualization backend, and the reason is because it needs to accept users' input on defining the **keyword dictionaries for grouping storage tags** (also, I do not see a very good way of visualizing bar charts using `chrome://tracing`).
>     (2) Because it requires users' input, the users must first look into the memory profiler logs to see what contributes most of the memory footprint, and this is why those logs are stored as `.csv` because they need to be first digested by the users.
>     (3) Current profiler API designs are more geared towards performance profiling, which in my opinion is different from storage profiling (e.g., in storage profiling, you do not really need the `pause()` and the `resume()` API call).
>   * Based on these, I decided to keep the current GPU memory profiler from the existing profiler APIs because I do not see a very good way of integrating them.

I think the MXNet profiler features are not limited just to profiling performance but also memory. The current profiler code provides data structures for counter that you can increment and it collects all the stats behind the scenes.

You can find examples for how to use the profiler for your custom use cases:
https://github.com/apache/incubator-mxnet/blob/master/src/profiler/storage_profiler.h
and here:
https://github.com/anirudh2290/mxnet/blob/memory_profiler_poc2/src/profiler/pool_memory_profiler.h
https://github.com/anirudh2290/mxnet/blob/memory_profiler_poc2/src/storage/pooled_storage_manager.h#L160

It seems like where your use case differs is you want to dump in a different format. You can add support for additional format in https://github.com/apache/incubator-mxnet/blob/master/src/profiler/aggregate_stats.cc#L46

and expose that additional format in frontend dumps API.

piyushghai commented:
@mxnet-label-bot Update [Profiler, pr-awaiting-response, pr-work-in-progress]

ArmageddonKnight commented:
Thanks to all for the valuable feedbacks. I opened [**another PR**](https://github.com/apache/incubator-mxnet/pull/17376) to address them. Closing this PR now.
